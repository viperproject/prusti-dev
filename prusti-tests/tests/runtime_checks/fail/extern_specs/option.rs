//@run: 101
use prusti_contracts::*;

#[extern_spec]
impl<T> std::option::Option<T> {
    #[pure]
    #[insert_runtime_check]
    #[ensures(matches!(*self, Some(_)) == result)]
    pub fn is_some(&self) -> bool;

    #[pure]
    #[insert_runtime_check]
    #[ensures(self.is_some() == !result)]
    pub fn is_none(&self) -> bool;

    #[trusted]
    #[insert_runtime_check]
    #[requires(self.is_some())]
    pub fn unwrap(self) -> T;
}

#[trusted]
fn main() {
    let x: Option<()> = None;
    // obviously panics! but with RT checks it should panic
    // because precondition check fails!
    x.unwrap();
}
