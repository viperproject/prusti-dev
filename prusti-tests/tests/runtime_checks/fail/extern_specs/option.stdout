check function prusti_pre_check_item_unwrap is performed
check function prusti_post_check_item_is_some is performed
